src/librustdoc/html/static/js/search.js

```diff
@@ -133,11 +124,337 @@ window.initSearch = function(rawSearchIndex) {
+     * This function parses the next query element until it finds `endChar`, calling `getNextElem`
+     * to collect each element.
+     *
+     * If there is no `endChar`, this function will implicitly stop at the end without raising an
+     * error.
+     *
+     * @param {ParsedQuery} query
+     * @param {ParserState} parserState
+     * @param {Array<QueryElement>} elems - This is where the new {QueryElement} will be added.
+     * @param {string} endChar            - This function will stop when it'll encounter this
+     *                                      character.
+     */
+    function getItemsBefore(query, parserState, elems, endChar) {
+        var turns = 0;
+        while (parserState.pos < parserState.length) {
+            var c = parserState.userQuery[parserState.pos];
+            if (c === endChar) {
+                break;
+            } else if (c === "," && endChar !== "" && turns > 0) {
+                parserState.pos += 1;
+                continue;
+            } else if (c === ":" && isPathStart(parserState)) {
+                throw new Error("Unexpected `::`: paths cannot start with `::`");
+            } else if (c === ":" || isEndCharacter(c)) {
+                var extra = "";
+                if (endChar === ">") {
+                    extra = "`<`";
+                } else if (endChar === "") {
+                    extra = "`->`";
+                }
+                throw new Error("Unexpected `" + c + "` after " + extra);
+            }
+            var posBefore = parserState.pos;
+            getNextElem(query, parserState, elems, endChar === ">");
+            turns += 1;
+            // This case can be encountered if `getNextElem` encounted a "stop character" right from
+            // the start. For example if you have `,,`. In this case, we simply move up the current
+            // position to continue the parsing.
+            if (posBefore === parserState.pos) {
+                parserState.pos += 1;
+            }
+        }
+        // We are either at the end of the string or on the `endChar`` character, let's move forward
+        // in any case.
+        parserState.pos += 1;
+    }
+
+    /**
+     * Parses the provided `query` input to fill `parserState`. If it encounters an error while
+     * parsing `query`, it'll throw an error.
+     *
+     * @param {ParsedQuery} query
+     * @param {ParserState} parserState
+     */
+    function parseInput(query, parserState) {
+        var c, before;
+
+        while (parserState.pos < parserState.length) {
+            c = parserState.userQuery[parserState.pos];
+            if (isStopCharacter(c)) {
+                if (c === "," || c === " ") {
+                    parserState.pos += 1;
+                    continue;
+                } else if (c === "-") {
+                    if (isReturnArrow(parserState)) {
+                        break;
+                    }
+                    throw new Error("Unexpected `-` (did you mean `->`?)");
```

Comment:
This is great. There's a similar problem if you end a path with a lone `>`.

```
$ node src/tools/rustdoc-js/tester.js \
        --doc-folder build/x86_64-unknown-linux-gnu/doc/ \
        --crate-name std \
        --dump 'a>b' \
        --resource-suffix 1.61.0
{
    "original": "a>b",
    "userQuery": "a>b",
    "typeFilter": -1,
    "elems": [
        {
            "name": "a",
            "fullPath": [
                "a"
            ],
            "pathWithoutLast": [],
            "pathLast": "a",
            "generics": []
        }
    ],
    "returned": [],
    "foundElems": 1,
    "literalSearch": false,
    "error": null
}
```

Reply:
Doing the same then! :)